### ISSUE TITLE
Segmentation faults

### ISSUE THREAD

**Issue by guv** [pixie-lang/pixie#165]
The following fails with a "Segmentation fault". (commit a1883f94c6c7d7d2dd55fe0e939518e90db9adf1)

```clojure
(def time (ffi-fn libc "time" [] CInt))

(defmacro timing [& body]
  `(let [start# (time),
         result# (do ~@body),
         end# (time)]
     (printf (str "Elapsed time: " (- end# start#)))
     result#))


(defn fib [n]
  (if (<= n 1)
    n
    (loop [n (dec n), f2 1, f1 0]
      (if (< n 1)
        f2
       (recur (dec n), (+ f2 f1), f2)))))

(timing (fib 30))
```

**Comment by heyLu** [comment]
As a clarification, it also fails with expressions instead of `(fib 30)` and sometimes it doesn't fail at all. (`(timing (fib 30))` worked at least once for me).

Can you try in interpreted mode (`make run_interactive`)?

**Comment by guv** [comment]
In interpreted mode it runs just fine. I was not able to provoke the segmentation fault.

**Comment by guv** [comment]
It is also working when I load the source file into a running pixie-vm via ```load-file```.
Only running as ```./pixie-vm fib.pxi``` seems to be broken.

**Comment by halgari** [comment]
Writing correct ffi code is rather error-prone. In this case the call to time is missing an argument, I think this was causing the segfault. An easy way to avoid such problems is to use ffi-infer to import the c function: 

```clojure

(ns time-fib
    (require pixie.ffi-infer :as f))

(f/with-config {:includes ["time.h"]
                :library "c" }
               (f/defcfn time))

(defmacro timing [& body]
  `(let [start# (time nil),
         result# (do ~@body),
         end# (time nil)]
     (println (str "Elapsed time: " (- end# start#)))
     result#))


(defn fib [n]
  (if (<= n 1)
    n
    (loop [n (dec n), f2 1, f1 0]
      (if (< n 1)
        f2
       (recur (dec n), (+ f2 f1), f2)))))

(timing (fib 30))

```

This code no longer segfaults on my machine, and in addition, calling time without a arg results in an error.

**Comment by guv** [comment]
Thank you. I tried that implementation and got the following exception.

```
Error:  in internal function load-reader

in pixie function with-config

in /data/pixie-dev/pixie/pixie/ffi-infer.pxi at 128:23
            *library* (ffi-library (str "lib" (:library config) "." pixie.platform/so-ext))]
                      ^
in internal function ffi-library

RuntimeException: /usr/lib/x86_64-linux-gnu/libc.so: invalid ELF header
```

I think that kind of exception has been the reason why I chose ```ffi-fn``` instead of ```defcfn``` .
Examples of both of these can be found in the codebase - so I was not sure which one to use.

**Comment by heyLu** [comment]
@guv I tried to fix that in #173, but I get the following error then:

```
c++ -m64 -std=c++11 /tmp/tmp.cpp -I/home/lu/t/pixie/pixie  -o /tmp/a.out && /tmp/a.out
Error:  in internal function load-reader

in pixie function main

RuntimeException: First argument must be an ExternalLib
```

I'm not sure why, but it might be related to #172, because regrettably I only I tested #173 before #172 was merged.

**Comment by guv** [comment]
I tried the previous code again with commit bbc22db6f2f8218e7833d7df5c7556130ca20be3 .
It works now. Though it should be added that it needs boost to be installed as runtime dependency.
